### PR TITLE
Implement scroll/scrollBy/scrollIntoView API methods

### PR DESCRIPTION
**Pull request by AlexKamaev**
API:

```await t.scroll(100, 500);``` - scrolls document to position left: 100, top: 500
```await t.scroll(target, 100, 500);``` - scrolls target element to position 100, 500 - inherits from VisibleElementAutomation
```await t.scroll('top | left | right | bottom | topLeft | topRight | bottomLeft | bottomRight| center');``` - scrolls document to predefined position
```await t.scroll(target, 'top');``` - scrolls target element to predefined position
```await t.scrollBy(1, 2);``` - scrolls document by 1, 2 from the current position
```await t.scrollBy(-1, -2);``` - scrolls document by decreasing its current position by 1, 2
```await t.scrollBy(target, 1, 2);``` - scrolls target element by 1, 2 from the current position
```await t.scrollBy(target, -1, -2);``` - scrolls target element by decreasing its current position by 1, 2
```await t.scrollIntoView(target);``` - scrolls all parent containers to get target element into view, works as `hover` but without mouse events and cursor
```await t.scroll(target, 100, 200, { offsetX: 1, offsetY: 1 });``` - scrolls element but use specified offset for locating this element
```await t.scroll(target, 'top', { offsetX: 1, offsetY: 1 }); ``` - same options added
```await t.scrollBy(target, 100, 200, { offsetX: -1, offsetY: -1 });``` - same options added
```await t.scrollIntoView(target, { offsetX: 1, offsetY: 1 });``` - same options added

### CODE REVIEW

File: src/api/test-controller/index.js [hunk 1]
```diff
@@ -209,6 +212,51 @@ export default class TestController {
...
     }
 
+    _parseSelectorForScroll (args) {
+        const selector = typeof args[0] === 'string' || isSelector(args[0]) ? args[0] : null;
+
+        if (selector)
+            args.shift();
+        else
+            // eslint-disable-next-line no-undef
+            return () => document.scrollingElement || document.documentElement;
```
Comment (miherlosev) [inline src/api/test-controller/index.js:222]: For which cases do we need the `document.scrollingElement` expression?

File: src/api/test-controller/index.js [hunk 2]
```diff
@@ -209,6 +212,51 @@ export default class TestController {
...
+
+        return selector;
+    }
+
+    _scroll$ (...args) {
+        let position = null;
+
+        if (args.length === 1 && typeof args[0] === 'string')
+            position = args.shift();
```
Comment (miherlosev) [inline src/api/test-controller/index.js:231]: Pack to the `_getPosition` function.

File: src/api/test-controller/index.js [hunk 3]
```diff
@@ -209,6 +212,51 @@ export default class TestController {
         return this._enqueueCommand('dragToElement', DragToElementCommand, { selector, destinationSelector, options });
     }
 
+    _parseSelectorForScroll (args) {
```
Comment (miherlosev) [inline src/api/test-controller/index.js:215]: You retrieve the selector from arguments, so the `_getSelectorForScroll` is a more suitable name.

File: src/client/automation/playback/set-scroll.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
+import hammerhead from '../deps/hammerhead';
```
Comment (miherlosev) [inline src/client/automation/playback/set-scroll.js:1]: Rewrite the new files with TypeScript.

**Review by miherlosev — CHANGES_REQUESTED** [review]
In general, it's good. Just a few refactoring related issues.

File: src/api/test-controller/index.js (hunk 1 above)
Comment (AlexKamaev) [reply src/api/test-controller/index.js:222]: Safari does not work correctly with document.documentElement for some reason. I checked it on BrowserStack MacOS machines and real iPhone device

File: src/api/test-controller/index.js (hunk 2 above)
Comment (AlexKamaev) [reply src/api/test-controller/index.js:231]: agree

File: src/api/test-controller/index.js (hunk 3 above)
Comment (AlexKamaev) [reply src/api/test-controller/index.js:215]: agree

File: src/client/automation/playback/set-scroll.js (hunk 4 above)
Comment (AlexKamaev) [reply src/client/automation/playback/set-scroll.js:1]: ok

File: src/api/test-controller/index.js (hunk 1 above)
Comment (miherlosev) [reply src/api/test-controller/index.js:222]: Add comment please for that.

**Review by miherlosev — APPROVED** [review]

**Comment by AlexKamaev** [comment]
@testcafe-build-bot

**Comment by DIRECTcut** [comment]
Names are a bit misleading.

We use `scroll` to scroll **to** a position,
and we use `scrollBy` to scroll **by** a certain number of pixels.

I propose that we rename the methods like so:

`scroll` -> `scrollTo`
`scrollBy` -> `scroll`

**Review by DIRECTcut — CHANGES_REQUESTED** [review]
Names are a bit misleading.

We use `scroll` to scroll **to** a position,
and we use `scrollBy` to scroll **by** a certain number of pixels.

I propose that we rename the methods like so:

`scroll` -> `scrollTo`
`scrollBy` -> `scroll`

**Review by AndreyBelym — APPROVED** [review]

**Review by miherlosev — COMMENTED** [review]
We are using naming from the similar element methods
Element.scroll - https://developer.mozilla.org/en-US/docs/Web/API/Element/scroll
Element.scrollBy - https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollBy

**Review by DIRECTcut — APPROVED** [review]
Works for me

**Pull request merged**